zcash_client_sqlite/src/wallet.rs

```diff
@@ -2048,4 +2049,138 @@ mod tests {
+        check_balance(&st, 1, NonNegativeAmount::ZERO);
+
+        // Create a fake transparent output.
+        let value = NonNegativeAmount::from_u64(100000).unwrap();
+        let outpoint = OutPoint::new([1u8; 32], 1);
+        let txout = TxOut {
+            value: value.into(),
+            script_pubkey: taddr.script(),
+        };
+
+        // Pretend the output was received in the chain tip.
+        let height = st.wallet().chain_height().unwrap().unwrap();
+        let utxo =
+            WalletTransparentOutput::from_parts(outpoint.clone(), txout.clone(), height).unwrap();
+        st.wallet_mut()
+            .put_received_transparent_utxo(&utxo)
+            .unwrap();
+
+        // The wallet should detect the balance as having 1 confirmation.
+        check_balance(&st, 0, value);
+        check_balance(&st, 1, value);
+        check_balance(&st, 2, NonNegativeAmount::ZERO);
+
+        // Shield the output.
+        let input_selector = GreedyInputSelector::new(
+            fixed::SingleOutputChangeStrategy::new(FixedFeeRule::non_standard(Amount::zero())),
+            DustOutputPolicy::default(),
+        );
+        let txid = st
+            .shield_transparent_funds(
+                &input_selector,
+                value,
+                &usk,
+                &[*taddr],
+                &MemoBytes::empty(),
+                NonZeroU32::new(1).unwrap(),
+            )
+            .unwrap();
+
+        // The wallet should have zero transparent balance, because the shielding
+        // transaction can be mined.
+        check_balance(&st, 0, NonNegativeAmount::ZERO);
+        check_balance(&st, 1, NonNegativeAmount::ZERO);
+        check_balance(&st, 2, NonNegativeAmount::ZERO);
+
+        // Mine the shielding transaction.
+        let (mined_height, _) = st.generate_next_block_including(txid);
+        st.scan_cached_blocks(mined_height, 1);
+
+        // The wallet should still have zero transparent balance.
+        check_balance(&st, 0, NonNegativeAmount::ZERO);
+        check_balance(&st, 1, NonNegativeAmount::ZERO);
+        check_balance(&st, 2, NonNegativeAmount::ZERO);
+
+        // Unmine the shielding transaction via a reorg.
+        st.wallet_mut()
+            .truncate_to_height(mined_height - 1)
+            .unwrap();
+
+        // The wallet should still have zero transparent balance.
+        check_balance(&st, 0, NonNegativeAmount::ZERO);
+        check_balance(&st, 1, NonNegativeAmount::ZERO);
+        check_balance(&st, 2, NonNegativeAmount::ZERO);
+
+        // Expire the shielding transaction.
+        let expiry_height = st.wallet().get_transaction(txid).unwrap().expiry_height();
+        st.wallet_mut().update_chain_tip(expiry_height).unwrap();
+
+        // The transparent output should be spendable again, with more confirmations.
+        check_balance(&st, 0, value);
+        check_balance(&st, 1, value);
+        check_balance(&st, 2, value);
```

Comment:
I added this to the test because it occurred to me that this is something that can happen to user wallets, particularly for SDK 2.0.0 where transactions will be using a fixed fee. However, this part of the test is failing because we don't have any handling for this situation.

Reply:
Let's talk tomorrow to figure out where it makes sense to handle this. Up until now transaction expiry has been treated as sort of a passive thing. I think that the correct solution here is going to be for the queries that are computing transparent balance to ensure that the spending transactions have not expired unmined. So, a UTXO might have a "spent in tx" field that is non-null, but that UTXO should still be treated as spendable if the referenced transaction has expired unmined, and future spends of that UTXO should then overwrite the spent-in-tx field.

Reply:
I suspect this should be a post-2.0 cleanup.

Reply:
Opened #986.